File: commands/Misc/mimic.js
const { Command } = require("klasa");
const { clean } = require("../../utils/utils.js");

class Mimic extends Command {
  constructor(...args) {
    super(...args, {
      description: "Copy someone and talk as them.",
      aliases: ["copycat"],
      requiredPermissions: ["MANAGE_WEBHOOKS"],
      usage: "<member:member|user:user> <message:...string>",
      usageDelim: " ",
      runIn: ["text"],
      permissionLevel: 4
    });
  }

  async run(msg, [user, message]) {
    if(msg.deletable) await msg.delete().catch(() => null);
    const name = user.displayName || user.username;
    const avatar = user.user ? user.user.displayAvatarURL({ format: "png", size: 2048 }) : user.displayAvatarURL({ format: "png", size: 2048 });
    const webhook = await msg.channel.createWebhook(name, { avatar });
    await webhook.send(clean(msg, message));
    await webhook.delete();
  }
}

module.exports = Mimic;

File: commands/misc/mimic.go
package misc

import (
  "github.com/sapphire-cord/sapphire"
  "github.com/bwmarrin/discordgo"
)

// Copy someone and talk as them.
// Usage: <@user> <message:string...>
// Aliases: copycat
// guild only
func Mimic(ctx *sapphire.CommandContext) {
  user := ctx.Arg(0).AsUser()
  name := user.Username
  // Try to find it as a member and use nicknames if available
  member := ctx.Member(user.ID)

  if member != nil && member.Nick != "" {
    name = member.Nick
  }

  w, err := ctx.Session.WebhookCreate(ctx.Channel.ID, name, user.AvatarURL("2048"))

  if err != nil {
    ctx.Reply("Failed to create a webhook, make sure i can **manage webhooks** in this channel.")
    return
  }

  ctx.Session.ChannelMessageDelete(ctx.Channel.ID, ctx.Message.ID)

  _, err = ctx.Session.WebhookExecute(w.ID, w.Token, false, &discordgo.WebhookParams{
    Content:ctx.JoinedArgs(1),
    AvatarURL: user.AvatarURL("2048"),
  })

  if err != nil {
    ctx.Error(err)
    return
  }

  ctx.Session.WebhookDelete(w.ID)
}
